python/lsst/ts/MTAirCompressor/aircompressor_csc.py

```diff
@@ -150,79 +107,156 @@ def add_arguments(cls, parser: argparse.ArgumentParser) -> None:
+        await super().close_tasks()
         if self.simulation_mode == 1:
             await self.simulator.shutdown()
             await self.simulator_future.cancel()
-        await super().close_tasks()
-
-    async def begin_start(self, data):
+        self.poll_task.cancel()
+        await self.disconnect()
+
+    async def log_modbus_error(self, modbus_error, msg="", ignore_timeouts=False):
+        if isinstance(modbus_error.exception, pymodbus.exceptions.ConnectionException):
+            await self.disconnect()
+
+        if not ignore_timeouts:
+            if self.summary_state != salobj.State.FAULT and (
+                self._failed_time is None
+                or monotonic() < self._failed_time + self.grace_period
+            ):
+                self.log.error(str(modbus_error))
+                if self._failed_time is None:
+                    self.log.warning(
+                        "Lost compressor connection, will try to reconnect for"
+                        f" {self.grace_period} seconds"
+                    )
+                    self._failed_time = monotonic()
+                return
+
+        try:
+            await self.fault(modbus_error.exception.original_code, msg)
+        except AttributeError:
+            if isinstance(
+                modbus_error.exception, pymodbus.exceptions.ConnectionException
+            ):
+                await self.fault(ErrorCode.COULD_NOT_CONNECT, msg + str(modbus_error))
+            else:
+                await self.fault(ErrorCode.MODBUS_ERROR, msg + str(modbus_error))
+
+        self._failed_time = None
+
+    async def connect(self):
+        if self.connection is None:
+            self.connection = ModbusClient(self.hostname, self.port)
+        ret = self.connection.connect()
+        if ret is False:
+            raise ModbusError(
+                pymodbus.exceptions.ConnectionException(
+                    f"Cannot establish connection to {self.hostname}:{self.port}"
+                )
+            )
+        if self.model is None:
+            self.model = MTAirCompressorModel(self.connection, self.unit)
+        await self.update_compressor_info()
+        self.log.debug(f"Connected to {self.hostname}:{self.port}")
+
+    async def disconnect(self):
+        self.model = None
+        if self.connection is not None:
+            self.connection.close()
+            self.connection = None
+
+    async def end_start(self, data):
         """Enables communication with the compressor."""
         if self.simulation_mode == 1:
             self.hostname = "localhost"
             self.unit = 1
 
             def run_sim():
+                # Returned ModbusTcpServer is subclass of socketserver.ThreaingTCPServer
+                # socketserver.ThreaingTCPServer stores address and host in
+                # server_address local variable.
                 self.simulator = create_server()
+                self.host, self.port = self.simulator.server_address
```

Comment:
(regarding line 192-195 blow): For a future update please consider having the simulator notify your CSC that it is properly running, so you can sleep until it is running instead of sleeping for 2 seconds.

One way to do this is to use `loop.call_soon_threadsafe` to set an asyncio.Future to done.

Reply:
Is there any example of that? I am not sure how that should work. Anyway, there is a simple solution: call create_simulator from the main thread, before starting only serve_forever in the executor.